File: python/src/binary.py
from typing import List


def binary_search(arr: List[int], target: int) -> int:
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = left + right
        if arr[mid] == target:
            return mid
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

def firstOccurence(arr:List[int],target:int)->int:
        l,r=0,len(arr)-1
        ans= -1
        while l<=r:
            mid = (l+r)
            if arr[mid] == target:
                ans=mid
                r=mid-1
            elif arr[mid]<target:
                l=mid+1
            else:
                r=mid-1
    

 __name__ == "__main__":
    arr = [int(x) for x in input().split()]
    target = int(input())
    res = binary_search(arr, target)
 print(res)

File: rust/src/binary.rs
use std::io;

fn binary_search(arr: &[i32], target: i32) -> i32 {
    let mut left = 0;
    let mut right = arr.len() as i32 - 1;
    while left <= right {
        let mid = (left + right) / 2;
        println!("compared {:?} and {:?}(mid)", target, mid);
        if arr[mid as usize] == target {
            println!("found");
            return mid;
        }
        if arr[mid as usize] < target {
            println!("target is bigger");
            left = mid + 1;
        } else {
            println!("target is smaller");
            right = mid - 1;
        }
    }
    -1
}

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("failed to read line");
    let arr: Vec<i32> = input
        .trim()
        .split_whitespace()
        .map(|x| x.parse().expect("Not a number"))
        .collect();
    println!("{:?}", arr);
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("failed to read line");
    let target: i32 = input.trim().parse().expect("not a number");
    let res = binary_search(&arr, target);
    println!("{}", res)
}
